translator/util/ec2util/ec2util.go

```diff
@@ -43,6 +46,14 @@ func initEC2UtilSingleton() (newInstance *ec2Util) {
 		return
 	}
 	md := ec2metadata.New(ses)
+	for i := 0; i < allowedRetries; i++ {
```

Comment:
My concern with this fix is 
1. how do we know 3 retries is enough for network getting online for those customers used to have this issue? Is 3 seconds long enough?
2. Will this 3 seconds(or potentially longer if we figured we need to wait more) delay for on-prem customers break any experience? Like on-prem customer missing 3 secs of metrics, which could be valuable during a launch.

Reply:
Maybe this can be configurable.

Reply:
I considered making the retries configurable. I'm not opposed to it. 3 was an arbitrary number. Originally I tested with 10 as the retries. My push back would be if the EC2 instance takes longer than 3 seconds for network activity then there are probably other issues going on with the host/OS. 

With the on-prem impact, I think that's a valid concern. But there is already a way for users to mitigate that by specifying on-prem mode themselves, which short circuits before we try to connect to the EC2 metadata endpoint. I don't think it's ideal, but currently not sure how else to implement this that would not require any change for people running on-prem and using auto detection.

Reply:
If I'm reading the right code at https://github.com/aws/amazon-cloudwatch-agent/blob/master/translator/util/sdkutil.go#L35 customer can config in EC2 and short circuits, but no on-prem, I didn't find anywhere a onPrem flag can bypass this method DetectAgentMode with my quick search.

Reply:
Let me dig into that a little more. I was really only looking at 
```go
	if context.CurrentContext().Mode() == config.ModeOnPrem {
		return
	}
```
inside of the `initEC2UtilSingleton` function.

Reply:
Looking at where it gets called, in the translator, 
```go
var inputMode = flag.String("mode", "ec2", "Please provide the mode, i.e. ec2, onPrem")
```
The default value for the flag in the translator is `ec2`. Then in the `DetectAgentMode` function,
```go
if configuredMode != "auto" {
    return configuredMode
}
```
if the input mode is not `auto`, then it immediately returns that string. 

The `start-amazon-cloudwatch-agent` binary runs the translator with the mode set to auto ☹️ so you're right. 

Personally, I think that the simplest way to resolve this is to extend the if statement to also check for a negative (non-empty) value from the environment if it is running in AWS 🤔 
```go
if runInAws == config.RUN_IN_AWS_TRUE {
    fmt.Println("I! Detected from ENV instance is EC2")
    return config.ModeEC2
} else if runInAws == config.RUN_IN_AWS_FALSE {
    fmt.Println("I! Detected from ENV instance is OnPrem")
    return config.ModeOnPrem
}
```

Thoughts?

Reply:
Currently, not-in-aws == onPrem, as long as this does not default to false, that should works.

Reply:
`runInAws` is derived from the environment, and `os.Getenv()` defaults to the empty string. If we introduced a const like:
```go
RUN_IN_AWS_FALSE = "False"
```
we could check against a non-empty value, so that wouldn't impact existing configuration. Would still love to hear any suggestions for how to account for this without impacting existing users who are on-prem and don't want to modify their environment, though - I still haven't thought of how to.

Reply:
Going to update this. Pivoting from the implementation here that introduces a rudimentary retry when the network is not available and we try to check for IMDS, because @ZhenyuTan-amz is right and we shouldn't add the retry for the case where a host should not interact with IMDS because it's not hosted on AWS.

New approach is to use the `net` package to iterate over the network interfaces on the host, and block until there is a non-loopback network interface that is `up`. This should help safeguard (at least a little bit) for the scenario where the host is on-prem, and doesn't need to check the IMDS endpoint.

Reply:
Tested it on my MacBook (on-prem) and saw that it did not attempt to sleep after finding active network access.
```
3c22fbe51e1b:bin andhuy$ sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m auto -c file:/opt/aws/amazon-cloudwatch-agent/bin/config.json -s
/opt/aws/amazon-cloudwatch-agent/bin/config-downloader --output-dir /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d --download-source file:/opt/aws/amazon-cloudwatch-agent/bin/config.json --mode auto --config /opt/aws/amazon-cloudwatch-agent/etc/common-config.toml --multi-config default
2022/03/09 09:49:24 Name: lo0, Flags: up|loopback|multicast
2022/03/09 09:49:24 Name: gif0, Flags: pointtopoint|multicast
2022/03/09 09:49:24 Name: stf0, Flags: 0
2022/03/09 09:49:24 Name: en6, Flags: up|broadcast|multicast
2022/03/09 09:49:24 E! ec2metadata is not available
I! Detected the instance is OnPrem
Got Home directory: /var/root
I! Set home dir Linux: /var/root
I! SDKRegionWithCredsMap region:  us-east-1
Successfully fetched the config and saved in /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_config.json.tmp
/opt/aws/amazon-cloudwatch-agent/bin/config-translator --input /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json --input-dir /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d --output /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.toml --mode auto --config /opt/aws/amazon-cloudwatch-agent/etc/common-config.toml --multi-config default
2022/03/09 09:49:24 Name: lo0, Flags: up|loopback|multicast
2022/03/09 09:49:24 Name: gif0, Flags: pointtopoint|multicast
2022/03/09 09:49:24 Name: stf0, Flags: 0
2022/03/09 09:49:24 Name: en6, Flags: up|broadcast|multicast
2022/03/09 09:49:25 E! ec2metadata is not available
I! Detected the instance is OnPrem
2022/03/09 09:49:25 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_config.json.tmp ...
Valid Json input schema.
Got Home directory: /var/root
I! Set home dir Linux: /var/root
I! SDKRegionWithCredsMap region:  us-east-1
No csm configuration found.
Under path : /logs/ | Info : Got hostname 3c22fbe51e1b.ant.amazon.com as log_stream_name
Configuration validation first phase succeeded
/opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent -schematest -config /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.toml
Configuration validation second phase succeeded
Configuration validation succeeded
```